### PR TITLE
Add `with` adaptor for maps with parseable keys

### PR DESCRIPTION
**Pull request by xlambein**
This is an implementation of https://github.com/automerge/autosurgeon/issues/19.

Note that I also added a tweak to the derive macros to allow a path in `#[autosurgeon(with = ...)]`, which before only allowed identifiers.  This makes the use of the new feature easier.

### CODE REVIEW

**Comment by alexjg** [comment]
I had a few thoughts on naming and organisation when reading through this, I put in a PR for your branch. Let me know what you think.

**Comment by xlambein** [comment]
Thanks for the changes! :-) They're merged into the PR

File: autosurgeon/src/hydrate.rs [hunk 1]
```diff
@@ -188,6 +188,8 @@ pub enum HydrateError {
     Automerge(#[from] automerge::AutomergeError),
     #[error("unexpected {0}")]
     Unexpected(Unexpected),
+    #[error("parse error: {0}")]
+    Parse(Box<dyn std::error::Error>),
```
Comment (alexjg) [inline autosurgeon/src/hydrate.rs:192]: Apologies for not spotting it earlier but I think we should probably add a `Send + Sync` bound here, it's reasonably likely that people will want to pass these errors around in a multithreaded context. We should probably also name the variant `ParseMapKey`.

Comment (xlambein) [reply autosurgeon/src/hydrate.rs]: Done :]

**Comment by alexjg** [comment]
This is now published in `0.5.0`

**Comment by xlambein** [comment]
Thanks for your help! :-)

**Pull request merged**